From 328878255858cdbe3fcd6574fa47c23c49990671 Mon Sep 17 00:00:00 2001
From: Benjamin Walsh <benjamin.walsh@windriver.com>
Date: Wed, 24 Mar 2010 16:20:57 -0400
Subject: [PATCH] WR Linux 3.0.3: Octeon/kexec: fix USB breakage

When rebooting via kexec, the USB subsystem is reset. However, if
the USB subsystem is not present, such as on 38xx/58xx boards,
a build error is seen if configuring kexec in the image.

Signed-off-by: Benjamin Walsh <benjamin.walsh@windriver.com>
---
 arch/mips/cavium-octeon/hal.c             |   29 +++++++++++++---------
 arch/mips/cavium-octeon/octeon-usb-host.c |   38 ++++++++++++++++++++++++----
 2 files changed, 49 insertions(+), 18 deletions(-)

diff --git a/arch/mips/cavium-octeon/hal.c b/arch/mips/cavium-octeon/hal.c
index ce9bff4..a617ccc 100644
--- a/arch/mips/cavium-octeon/hal.c
+++ b/arch/mips/cavium-octeon/hal.c
@@ -768,10 +768,18 @@ static void octeon_kexec_shutdown_other_cpus(void *ignore)
 }
 #endif /* CONFIG_SMP */
 
-/* externs for USB shutdown */
-#define MAX_USB_PORTS   10
-extern struct platform_device *pdev_glob[MAX_USB_PORTS];
-extern void platform_device_unregister(struct platform_device *pdev);
+/* USB shutdown hooks, overriden in octeon_usb_module_init() if
+ * CAVIUM_OCTEON_USB is configured in or when module is loaded */
+static void __shutdown_usb_ports(void) { /* do nothing */ }
+static void (*kexec_octeon_shutdown_usb_ports)(void) = __shutdown_usb_ports;
+void kexec_octeon_register_shutdown_usb_ports(void(*fn)(void))
+{
+	kexec_octeon_shutdown_usb_ports = fn;
+}
+void kexec_octeon_unregister_shutdown_usb_ports(void)
+{
+	kexec_octeon_shutdown_usb_ports = __shutdown_usb_ports;
+}
 
 extern int cvmx_mgmt_port_shutdown(int port);
 
@@ -797,14 +805,7 @@ void octeon_kexec_shutdown(void)
 	}
 #endif /* CONFIG_SMP */
 
-	/* shutdown all USB ports */
-	int i=0;
-	for(i=0; i < MAX_USB_PORTS; i++) {
-		if(pdev_glob[i]) {
-			platform_device_unregister(pdev_glob[i]);
-			pdev_glob[i] = NULL;
-		}
-	}
+	kexec_octeon_shutdown_usb_ports();
 
 	/* shutdown all mgmt ports */
 	int port;
@@ -978,3 +979,7 @@ void octeon_crash_dump_setup(void)
 }
 #endif
 
+#ifdef CONFIG_KEXEC
+EXPORT_SYMBOL(kexec_octeon_register_shutdown_usb_ports);
+EXPORT_SYMBOL(kexec_octeon_unregister_shutdown_usb_ports);
+#endif
diff --git a/arch/mips/cavium-octeon/octeon-usb-host.c b/arch/mips/cavium-octeon/octeon-usb-host.c
index 82e4bc3..fee5143 100644
--- a/arch/mips/cavium-octeon/octeon-usb-host.c
+++ b/arch/mips/cavium-octeon/octeon-usb-host.c
@@ -718,6 +718,34 @@ static struct device_driver octeon_usb_driver = {
 
 #define MAX_USB_PORTS	10
 struct platform_device *pdev_glob[MAX_USB_PORTS];
+static void __shutdown_usb_ports(void)
+{
+	/* shutdown all USB ports */
+	int i=0;
+	for(i=0; i < MAX_USB_PORTS; i++) {
+		if(pdev_glob[i]) {
+			platform_device_unregister(pdev_glob[i]);
+			pdev_glob[i] = NULL;
+		}
+	}
+}
+
+#ifdef CONFIG_KEXEC
+void kexec_octeon_register_shutdown_usb_ports(void(*fn)(void));
+void kexec_octeon_unregister_shutdown_usb_ports(void);
+static void __register_for_kexec_shutdown(void)
+{
+	kexec_octeon_register_shutdown_usb_ports(__shutdown_usb_ports);
+}
+static void __unregister_for_kexec_shutdown(void)
+{
+	kexec_octeon_unregister_shutdown_usb_ports();
+}
+#else
+static void __register_for_kexec_shutdown(void) { /* do nothing */ }
+static void __unregister_for_kexec_shutdown(void) { /* do nothing */ }
+#endif
+
 static int __init octeon_usb_module_init(void)
 {
 	int num_devices = cvmx_usb_get_num_ports();
@@ -728,6 +756,8 @@ static int __init octeon_usb_module_init(void)
 		return -ENOMEM;
 	}
 
+	__register_for_kexec_shutdown();
+
 	printk(KERN_INFO "OcteonUSB: Detected %d ports\n", num_devices);
 	for (device = 0; device < num_devices; device++) {
 		struct resource irq_resource;
@@ -757,13 +787,9 @@ static int __init octeon_usb_module_init(void)
 
 static void __exit octeon_usb_module_cleanup(void)
 {
-	int i;
 	DEBUG_CALL("OcteonUSB: %s called\n", __func__);
-	for (i = 0; i < MAX_USB_PORTS; i++)
-		if (pdev_glob[i]) {
-			platform_device_unregister(pdev_glob[i]);
-			pdev_glob[i] = NULL;
-		}
+	__unregister_for_kexec_shutdown();
+	__shutdown_usb_ports();
 	driver_unregister(&octeon_usb_driver);
 }
 
-- 
1.6.0.4

